File: 2002PDC/WordCount/VB/WordCount.cs
/*=====================================================================
  File:      WordCount.cs

  Summary:   Demonstrates how to create a WordCount application
             using various COM+ 2.0 library types.

---------------------------------------------------------------------
  This file is part of the Microsoft COM+ 2.0 SDK Code Samples.

  Copyright (C) 2000 Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.

THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/

// Add the classes in the following namespaces to our namespace
using System;
using System.IO;
using System.Collections;

namespace CompWordCounter{

///////////////////////////////////////////////////////////////////////////////


// The WordCounter class
public class WordCounter {
    public WordCounter() { /* No interesting construction */ }

    // Each object of this class keeps a running total of the files its processed
    // The following members hold this running information
    Int64 totalLines = 0; 
    Int64 totalWords = 0;
    Int64 totalChars = 0;
    Int64 totalBytes = 0;

    // The set of all words seen (sorted alphabetically)
    SortedList wordCounter = new SortedList();

    // The following methods return the running-total info
    public Int64 TotalLines { get { return totalLines; } }
    public Int64 TotalWords { get { return totalWords; } }
    public Int64 TotalChars { get { return totalChars; } }
    public Int64 TotalBytes { get { return totalBytes; } }

    // This method calculates the statistics for a single file.
    // This file's info is returned via the out parameters
    // The running total of all files is maintained in the data members
    public Boolean CountStats(String pathname, 
        out Int64 numLines, out Int64 numWords, out Int64 numChars, out Int64 numBytes) {

        Boolean Ok = true;  // Assume success
        numLines = numWords = numChars = numBytes = 0;  // Initialize out params to zero
        try {
            // Attempt to open the input file for read-only access
            FileStream fsIn = new FileStream(pathname, FileMode.Open, FileAccess.Read, FileShare.Read);
            numBytes = fsIn.Length;
            StreamReader sr = new StreamReader(fsIn);

            // Process every line in the file
            for (String Line = sr.ReadLine(); Line != null; Line = sr.ReadLine()) {
                numLines++;
                numChars += Line.Length;
                String[] Words = Line.Split(null);  // Split the line into words
                numWords += Words.Length;
                for (int Word = 0; Word < Words.Length; Word++) {
                    if (Words[Word] != null) {
                        if (!wordCounter.ContainsKey(Words[Word])) {
                            // If we've never seen this word before, add it to the sorted list with a count of 1
                            wordCounter.Add(Words[Word], 1);
                        } else {
                            // If we have seen this word before, just increment its count
                            wordCounter[Words[Word]] = (Int32) wordCounter[Words[Word]] + 1;
                        }
                    }
                }
            }
            // Explicitly close the StreamReader to properly flush all buffers
            sr.Close(); // This also closes the FileStream (fsIn)
        }
        catch (FileNotFoundException) {
            // The specified input file could not be opened
            Ok = false;
        }
        // Increment the running totals with whatever was discovered about this file
        totalLines += numLines;
        totalWords += numWords;
        totalChars += numChars;
        totalBytes += numBytes;
        return(Ok);
    }

    // Returns an enumerator for the words (sorted alphabetically)
    public IDictionaryEnumerator GetWordsAlphabeticallyEnumerator() {
        return (IDictionaryEnumerator) wordCounter.GetEnumerator();
    }


    // This nested class is only used to sort the words by occurrance
    // An instance of this class is created for each word
    public class WordOccurrance : IComparable {

        // Members indicating the number of times this word occurred and the word itself
        private int occurrances;
        private String word;

        // Constructor
        public WordOccurrance(int occurrances, String word) {
            this.occurrances = occurrances;
            this.word = word;
        }

        // Sorts two WordOccurrance objects by occurrance first, then by word
        public int CompareTo(Object o) {
            // Compare the occurance of the two objects
            int n = occurrances - ((WordOccurrance)o).occurrances;
            if (n == 0) {
                // Both objects have the same ccurrance, sort alphabetically by word
                n = String.Compare(word, ((WordOccurrance)o).word);
            }
            return(n);
        }

        // Return the occurrance value of this word
        public int Occurrances { get { return occurrances; } }

        // Return this word
        public String Word { get { return word; } }
    }


    // Returns an enumerator for the words (sorted by occurrance)
    public IDictionaryEnumerator GetWordsByOccurranceEnumerator() {
        // To create a list of words sorted by occurrance, we need another SortedList object
        SortedList sl = new SortedList();

        // Now, we'll iterate through the words alphabetically
        IDictionaryEnumerator de = GetWordsAlphabeticallyEnumerator();
        while (de.MoveNext()) {

            // For each word, we create a new WordOccurrance object which
            // contains the word and its occurrance value.
            // The WordOccurrance class contains a CompareTo method which knows
            // to sort WordOccurrance objects by occurrance value and then alphabetically by the word itself.
            sl.Add(new WordOccurrance((int)de.Value, (string)de.Key), null);
        }
        // Return an enumerator for the words (sorted by occurrance)
        return (IDictionaryEnumerator) sl.GetEnumerator();        
    }

    // Returns the number of unique words processed
    public int UniqueWords {
        get { return wordCounter.Count; }
    }
}


///////////////////////////////////////////////////////////////////////////////


}

///////////////////////////////// End of File /////////////////////////////////

File: 2002PDC/WordCount/VB/WordCount.vb

' Add the classes in the following namespaces to our namespace
imports System
imports System.IO
imports System.Collections
imports CompWordCounter

Option Explicit
Option Strict Off

' ///////////////////////////////////////////////////////////////////////////////


Public Module modmain

   Sub Main()
        ' Parse the command-line arguments
        dim ap as New WordCountArgParser
        dim args as string()



        args = Environment.GetCommandLineArgs()

        if not ap.Parse() then 
            ' An error occurrend while parsing
            exit sub
        end if


        ' If an output file was specified on the command-line, use it
        dim fsOut as FileStream  = nothing
        dim sw as StreamWriter = nothing
        if ap.OutputFile <> "" then
            fsOut = new FileStream(ap.OutputFile, FileMode.Create, FileAccess.Write, FileShare.None)
            sw = new StreamWriter(fsOut)

            ' By associating the StreamWriter with the console, the rest of 
            ' the code can think it's writing to the console but the console 
            ' object redirects the output to the StreamWriter
            Console.SetOut(sw)
        end if

        ' Create a WordCounter object to keep running statistics
        dim wc as New WordCounter

        ' Write the table header
        Console.WriteLine("Lines  Words  Chars  Bytes  Pathname")

        ' Iterate over the user-specified files
        dim e as IEnumerator = ap.GetPathnameEnumerator()
        while (e.MoveNext())
            dim NumLines, NumWords, NumChars, NumBytes as Int64
            ' Calculate the words stats for this file
            wc.CountStats(e.Current, NumLines, NumWords, NumChars, NumBytes)

            ' Display the results
            dim StrArgs(5) as String
            StrArgs(0) = NumLines.ToString()
            StrArgs(1) = NumWords.ToString() 
            StrArgs(2) = NumChars.ToString() 
            StrArgs(3) = NumBytes.ToString() 
            StrArgs(4) = e.Current 
            
            Console.WriteLine(System.String.Format("{0,5}  {1,5}  {2,5}  {3,5}  {4,5}", StrArgs))
        End While


        ' Done processing all files, show the totals
        Console.WriteLine("-----  -----  -----  -----  ---------------------")
        Console.WriteLine(System.String.Format("{0,5}  {1,5}  {2,5}  {3,5}  Total in all files", wc.TotalLines, wc.TotalWords, wc.TotalChars, wc.TotalBytes))

        ' If the user wants to see the word usage alphabetically, show it
        if ap.ShowAlphabeticalWordUsage then
            dim de as IDictionaryEnumerator = wc.GetWordsAlphabeticallyEnumerator()
            Console.WriteLine(System.String.Format("Word usage sorted alphabetically ({0} unique words)", wc.UniqueWords))
            while de.MoveNext()
                Console.WriteLine(System.String.Format("{0,5}: {1}", de.Value, de.Key))
            End While
        end if

        ' If the user wants to see the word usage by occurrance, show it
        if ap.ShowOccurranceWordUsage then
            dim de as IDictionaryEnumerator = wc.GetWordsByOccurranceEnumerator()
            Console.WriteLine(System.String.Format("Word usage sorted by occurrance ({0} unique words)", wc.UniqueWords))
            while de.MoveNext()
                Console.WriteLine(System.String.Format("{0,5}: {1}", de.Key.Occurrances, de.Key.Word))
            End While
        End If

        ' Explicitly close the console to guarantee that the StreamWriter object (sw) is flushed
        Console.Out.Close()
        if not (fsOut = nothing) then fsOut.Close()    

   End Sub

End Module


'///////////////////////////////// End of File /////////////////////////////////
